## Unión y depuración de la data

En este notebook uniremos toda la data recopilada anteriormente y crearemos un único DataFrame.  
Posteriormente realizaremos una limpieza de los datos.

In [1]:
import glob
import pandas as pd
from functools import reduce
import numpy as np
pd.options.mode.chained_assignment = None

En lo posterior, necesitamos que el notebook muestre todas las columnas del DataFrame:

In [2]:
pd.set_option('display.max_columns', None)

Unimos todos los csv para crear un único DataFrame:

In [3]:
dfLista = []
ruta = '/home/dsc/Python/Despoblacion/2.Data' 
csv_files = glob.glob(ruta + "/*11.csv")
csv_files.sort() 

for file in csv_files:
    df = pd.read_csv(file, dtype = {'Codigo CCAA' : str, 'Codigo Provincia' : str, 'Codigo Municipio' : str})
    dfLista.append(df) 
    
Despoblacion = reduce(lambda  left,right: pd.merge(left,right,on=['Codigo Municipio'],
                                            how='left'), dfLista)
Despoblacion.head()    

Nombre CCAA Codigo CCAA Nombre Provincia Codigo Provincia  Nombre Municipio  \
0  Pais Vasco          16      Araba/Álava               01  Alegría-Dulantzi   
1  Pais Vasco          16      Araba/Álava               01           Amurrio   
2  Pais Vasco          16      Araba/Álava               01             Añana   
3  Pais Vasco          16      Araba/Álava               01           Aramaio   
4  Pais Vasco          16      Araba/Álava               01           Armiñón   

  Codigo Municipio  Poblacion Total  Poblacion Hombre  Poblacion Mujer  \
0            01001           2803.0            1466.0           1337.0   
1            01002          10014.0            5030.0           4984.0   
2            01049            168.0              86.0             82.0   
3            01003           1496.0             784.0            712.0   
4            01006            229.0             112.0            117.0   

   Superficie Km2  Densidad Poblacion  Altitud Municipios  Nacimientos  \
0           19.95              140.50               568.0         37.0   
1           96.30              103.99               219.0        104.0   
2           21.92                7.66               574.0          1.0   
3           73.02               20.49               333.0         13.0   
4           12.97               17.66               467.0          2.0   

   Matrimonios  Defunciones  Total Regimen SS  Total Contratos  \
0         18.0         14.0             784.0              617   
1         37.0         88.0            4397.0             5861   
2          0.0          7.0              30.0               20   
3          3.0         10.0             224.0              141   
4          0.0          3.0             137.0              125   

   Contratos Hombres  Contratos Mujeres  Total Paro Registrado  \
0                453                164                   1992   
1               4843               1018                   8393   
2                 14                  6                     50   
3                 88                 53                    322   
4                 90                 35                    139   

   Parados Hombres  Parados Mujeres  Nº Establecimientos Agricultura  \
0              935             1057                              NaN   
1             3742             4651                              NaN   
2               16               34                              NaN   
3              177              145                              NaN   
4               60               79                              NaN   

   Nº Establecimientos Industria  Nº Establecimientos Construccion  \
0                            NaN                               NaN   
1                            NaN                               NaN   
2                            NaN                               NaN   
3                            NaN                               NaN   
4                            NaN                               NaN   

   Nº Establecimientos Servicios  Nº Centros Sanitarios  \
0                            NaN                    1.0   
1                            NaN                    2.0   
2                            NaN                    1.0   
3                            NaN                    1.0   
4                            NaN                    NaN   

   Nº Centros Educativos  Número de transacciones anuales 2011  \
0                    4.0                                  25.0   
1                    9.0                                 185.0   
2                    NaN                                   0.0   
3                    2.0                                   8.0   
4                    2.0                                   7.0   

   Nº Residencias Mayores  Nº Instituciones Religiosas y Militares  \
0                     1.0                                      1.0   
1                     3.0                                      NaN   
2              

In [4]:
Despoblacion.shape

(8116, 80)

Transformamos todos los valores del DataFrame en numéricos con dos decimales.  
Antes de ello, en la fase previa de recopilación de la data, se detectó que había tablas con variables que toman el valor \*.
Estos valores corresponden a variables extraídas del INE donde, por secreto estadístico, los registros que tienen un valor inferior a 5, les asigunan un \*.   
Localizamos las columnas que contienen \* y lo sustituimos por la media:

In [5]:
ColumnasAsterisco = Despoblacion.columns[Despoblacion.isin(['*']).any()]
ColumnasAsterisco

Index(['Edad < 16', '16 =< Edad =<64', 'Edad >= 65', 'Soltero/a', 'Casado/a',
       'Separado/a', 'Divorciado/a', 'Viudo/a', 'Nº de Jubilados',
       'Empresario que emplea personal', 'Empresario que no emplea personal',
       '1 hijo/a', '2 hijos/as', '3 hijos/as', '4 o más hijos/as',
       'Nº de Alquileres', 'Hogares Unipersonales', 'Hogares Familiares',
       '1 persona', '2 personas', '3 personas', '4 personas', '5 personas',
       '6 o más personas', 'Nº Hogares si Internet', 'Nº Hogares no Internet'],
      dtype='object')

In [6]:
Despoblacion[ColumnasAsterisco] = Despoblacion[ColumnasAsterisco].replace('*', 2)

Ahora si que ya podemos realizar el cambio de formato desde la columna número 4 en adelante:

In [7]:
Columnas = Despoblacion.columns[6 : ]
Despoblacion[Columnas] = Despoblacion[Columnas].astype(float).round(2)

In [8]:
Formato = Despoblacion.dtypes
print(Formato.to_string())

Nombre CCAA                                 object
Codigo CCAA                                 object
Nombre Provincia                            object
Codigo Provincia                            object
Nombre Municipio                            object
Codigo Municipio                            object
Poblacion Total                            float64
Poblacion Hombre                           float64
Poblacion Mujer                            float64
Superficie Km2                             float64
Densidad Poblacion                         float64
Altitud Municipios                         float64
Nacimientos                                float64
Matrimonios                                float64
Defunciones                                float64
Total Regimen SS                           float64
Total Contratos                            float64
Contratos Hombres                          float64
Contratos Mujeres                          float64
Total Paro Registrado          

In [9]:
Despoblacion.head()

Nombre CCAA Codigo CCAA Nombre Provincia Codigo Provincia  Nombre Municipio  \
0  Pais Vasco          16      Araba/Álava               01  Alegría-Dulantzi   
1  Pais Vasco          16      Araba/Álava               01           Amurrio   
2  Pais Vasco          16      Araba/Álava               01             Añana   
3  Pais Vasco          16      Araba/Álava               01           Aramaio   
4  Pais Vasco          16      Araba/Álava               01           Armiñón   

  Codigo Municipio  Poblacion Total  Poblacion Hombre  Poblacion Mujer  \
0            01001           2803.0            1466.0           1337.0   
1            01002          10014.0            5030.0           4984.0   
2            01049            168.0              86.0             82.0   
3            01003           1496.0             784.0            712.0   
4            01006            229.0             112.0            117.0   

   Superficie Km2  Densidad Poblacion  Altitud Municipios  Nacimientos  \
0           19.95              140.50               568.0         37.0   
1           96.30              103.99               219.0        104.0   
2           21.92                7.66               574.0          1.0   
3           73.02               20.49               333.0         13.0   
4           12.97               17.66               467.0          2.0   

   Matrimonios  Defunciones  Total Regimen SS  Total Contratos  \
0         18.0         14.0             784.0            617.0   
1         37.0         88.0            4397.0           5861.0   
2          0.0          7.0              30.0             20.0   
3          3.0         10.0             224.0            141.0   
4          0.0          3.0             137.0            125.0   

   Contratos Hombres  Contratos Mujeres  Total Paro Registrado  \
0              453.0              164.0                 1992.0   
1             4843.0             1018.0                 8393.0   
2               14.0                6.0                   50.0   
3               88.0               53.0                  322.0   
4               90.0               35.0                  139.0   

   Parados Hombres  Parados Mujeres  Nº Establecimientos Agricultura  \
0            935.0           1057.0                              NaN   
1           3742.0           4651.0                              NaN   
2             16.0             34.0                              NaN   
3            177.0            145.0                              NaN   
4             60.0             79.0                              NaN   

   Nº Establecimientos Industria  Nº Establecimientos Construccion  \
0                            NaN                               NaN   
1                            NaN                               NaN   
2                            NaN                               NaN   
3                            NaN                               NaN   
4                            NaN                               NaN   

   Nº Establecimientos Servicios  Nº Centros Sanitarios  \
0                            NaN                    1.0   
1                            NaN                    2.0   
2                            NaN                    1.0   
3                            NaN                    1.0   
4                            NaN                    NaN   

   Nº Centros Educativos  Número de transacciones anuales 2011  \
0                    4.0                                  25.0   
1                    9.0                                 185.0   
2                    NaN                                   0.0   
3                    2.0                                   8.0   
4                    2.0                                   7.0   

   Nº Residencias Mayores  Nº Instituciones Religiosas y Militares  \
0                     1.0                                      1.0   
1                     3.0                                      NaN   
2              

La dimensión del DataFrame es de 8.116 filas coincidiendo con el número de municipios en España en 2011.  
No obstante, veamos si hay algún Codigo Municipio duplicado:

In [10]:
Despoblacion[Despoblacion['Codigo Municipio'].duplicated()]

Empty DataFrame
Columns: [Nombre CCAA, Codigo CCAA, Nombre Provincia, Codigo Provincia, Nombre Municipio, Codigo Municipio, Poblacion Total, Poblacion Hombre, Poblacion Mujer, Superficie Km2, Densidad Poblacion, Altitud Municipios, Nacimientos, Matrimonios, Defunciones, Total Regimen SS, Total Contratos, Contratos Hombres, Contratos Mujeres, Total Paro Registrado, Parados Hombres, Parados Mujeres, Nº Establecimientos Agricultura, Nº Establecimientos Industria, Nº Establecimientos Construccion, Nº Establecimientos Servicios, Nº Centros Sanitarios, Nº Centros Educativos, Número de transacciones anuales 2011, Nº Residencias Mayores, Nº Instituciones Religiosas y Militares, Nº Hombres Extranjeros, Nº Mujeres Extranjeras, Nº Total Extranjeros, Nº Hombres Otra CCAA, Nº Mujeres Otra CCAA, Nº Total Otra CCAA, Nº edificios, Nº Inmuebles, Nº viviendas, Edad < 16, 16 =< Edad =<64, Edad >= 65, Soltero/a, Casado/a, Separado/a, Divorciado/a, Viudo/a, Sin Estudios, Con Estudios, Nº de Jubilados, Empresario que emplea personal, Empresario que no emplea personal, 1 hijo/a, 2 hijos/as, 3 hijos/as, 4 o más hijos/as, Nº de Alquileres, Hogares Unipersonales, Hogares Familiares, 1 persona, 2 personas, 3 personas, 4 personas, 5 personas, 6 o más personas, Edad Media Edificios, Uso Industrial, Uso Residelcial, Uso Terciario, Otros Usos, Nº Hogares si Internet, Nº Hogares no Internet, Nº Vehiculos, Nº Conductores, Deuda (Miles de €), Total impuestos (Miles de €), Superficie (Ha) Agricolas, Nº Explotaciones Agrícolas, Nº Explotaciones Ganaderas]
Index: []

Analizamos los valores nulos:

In [11]:
Valores = Despoblacion.isnull().sum()
print(Valores.to_string())

Nombre CCAA                                   0
Codigo CCAA                                   0
Nombre Provincia                              0
Codigo Provincia                              0
Nombre Municipio                              0
Codigo Municipio                              0
Poblacion Total                               0
Poblacion Hombre                              0
Poblacion Mujer                               0
Superficie Km2                                0
Densidad Poblacion                            0
Altitud Municipios                            0
Nacimientos                                   0
Matrimonios                                   0
Defunciones                                   0
Total Regimen SS                             36
Total Contratos                               0
Contratos Hombres                             0
Contratos Mujeres                             0
Total Paro Registrado                         0
Parados Hombres                         

**Total Regimen SS**: El campo contiene 36 valores nulos.  
La tabla de afiliados no contiene los 36 municipios correspondientes:

In [12]:
AfiliadosNulos = Despoblacion[Despoblacion['Total Regimen SS'].isnull()]
AfiliadosNulos = AfiliadosNulos['Codigo Municipio'].values.tolist()
AfiliadosNulos

['02901',
 '06902',
 '08904',
 '08903',
 '09102',
 '09478',
 '13903',
 '19031',
 '19116',
 '19171',
 '19182',
 '20907',
 '25912',
 '25913',
 '26035',
 '26077',
 '26181',
 '29902',
 '31119',
 '34110',
 '36901',
 '41903',
 '43907',
 '44142',
 '44203',
 '44204',
 '44211',
 '44218',
 '46902',
 '46903',
 '48914',
 '48915',
 '50042',
 '50047',
 '50080',
 '50902']

In [13]:
Afiliaciones11 = pd.read_csv('/home/dsc/Python/Despoblacion/2.Data/06_Output_Afiliados_SS_11.csv' ,
                             dtype = {'Codigo Municipio' : str})

Afiliaciones11[Afiliaciones11['Codigo Municipio'].isin(AfiliadosNulos)]

Empty DataFrame
Columns: [Codigo Municipio, Total Regimen SS]
Index: []

Eliminamos los 36 registros:

In [14]:
Despoblacion = Despoblacion.dropna(subset = ['Total Regimen SS'])

**Número de transacciones anuales 2011**: Casualmente son 4 municipios que posteriormente, en 2016, se fusionaron: Cercedo-Cotobade, Oza-Cesuras. Vemos si en el DataFrame de transacciones aparecen los municipios fusionados:


In [15]:
Despoblacion[Despoblacion['Número de transacciones anuales 2011'].isnull()].head()

Nombre CCAA Codigo CCAA Nombre Provincia Codigo Provincia  \
2145     Galicia          12        Coruña, A               15   
2182     Galicia          12        Coruña, A               15   
5280     Galicia          12       Pontevedra               36   
5281     Galicia          12       Pontevedra               36   

     Nombre Municipio Codigo Municipio  Poblacion Total  Poblacion Hombre  \
2145          Cesuras            15026           2225.0            1090.0   
2182     Oza dos Ríos            15063           3229.0            1622.0   
5280          Cerdedo            36011           2297.0            1063.0   
5281         Cotobade            36012           4432.0            2101.0   

      Poblacion Mujer  Superficie Km2  Densidad Poblacion  Altitud Municipios  \
2145           1135.0           79.46               28.00               231.0   
2182           1607.0           72.14               44.76               231.0   
5280           1234.0          134.70               17.05               374.0   
5281           2331.0           79.84               55.51               374.0   

      Nacimientos  Matrimonios  Defunciones  Total Regimen SS  \
2145          3.0          9.0         35.0             325.0   
2182         23.0         11.0         45.0             656.0   
5280          7.0          3.0         32.0             234.0   
5281         26.0         12.0         71.0             605.0   

      Total Contratos  Contratos Hombres  Contratos Mujeres  \
2145             67.0               40.0               27.0   
2182            392.0              286.0              106.0   
5280            154.0               64.0               90.0   
5281            272.0              170.0              102.0   

      Total Paro Registrado  Parados Hombres  Parados Mujeres  \
2145                 1396.0            940.0            456.0   
2182                 2512.0           1195.0           1317.0   
5280                 1319.0            670.0            649.0   
5281                 4957.0           2628.0           2329.0   

      Nº Establecimientos Agricultura  Nº Establecimientos Industria  \
2145                              NaN                            NaN   
2182                              NaN                            NaN   
5280                              NaN                            NaN   
5281                              NaN                            NaN   

      Nº Establecimientos Construccion  Nº Establecimientos Servicios  \
2145                               NaN                            NaN   
2182                               NaN                            NaN   
5280                               NaN                            NaN   
5281                               NaN                            NaN   

      Nº Centros Sanitarios  Nº Centros Educativos  \
2145                    1.0                    NaN   
2182                    1.0                    NaN   
5280                    1.0                    NaN   
5281                    2.0                    NaN   

      Número de transacciones anuales 2011  Nº Residencias Mayores  \
2145                                   NaN                     NaN   
2182                                   NaN                     NaN   
5280                                   NaN                     NaN   
5281                                   NaN                     NaN   

      Nº Instituciones Religiosas y Militares  Nº Hombres Extranjeros  \
2145                                      NaN                    21.0   
2182                                      NaN                    66.0   
5280                                      NaN                    75.0   
5281                                      NaN                    91.0   

      Nº Mujeres Extranjeras  Nº Total Extranjeros  Nº Hombres Otra CCAA  \
2145                    22.0                  43.0                  25.0   
2182                    61.0                 12

In [16]:
Transacciones11 = pd.read_csv('/home/dsc/Python/Despoblacion/2.Data/12_Output_Transacciones_Inmobiliarias_11.csv' ,
                              dtype = {'Codigo Provincia' : str, 'Codigo Municipio' : str})

In [17]:
Transacciones11[Transacciones11['Codigo Municipio'] == '36902']#Cercedo-Cotobade

Codigo Municipio  Número de transacciones anuales 2011
5344            36902                                   NaN

In [18]:
Transacciones11[Transacciones11['Codigo Municipio'] == '15902'] #Oza-Cesuras

Codigo Municipio  Número de transacciones anuales 2011
2218            15902                                  27.0

Eliminamos los registros de Cercedo y Cotobade.   
Tenemos el dato global de Oza-Cesuras, estimamos: la población es un 60/40, asignamos 16/11:

In [19]:
Despoblacion.loc[2145, 'Número de transacciones anuales 2011'] = 11
Despoblacion.loc[2183, 'Número de transacciones anuales 2011'] = 16

Despoblacion = Despoblacion.dropna(subset = ['Número de transacciones anuales 2011'])

**Nº Establecimientos Agricultura, Industria , Construccion , Servicios, Centros Sanitarios, Centros Educativos, Residencias Mayores, Instituciones Religiosas y Militares**: Los valores nulos corresponden a la falta de dichos centros en esos municipios.   
Los sustituimos por ceros:


In [20]:
ColumnasNan = ['Nº Establecimientos Agricultura', 'Nº Establecimientos Industria', 'Nº Establecimientos Construccion',
               'Nº Establecimientos Servicios', 'Nº Centros Sanitarios', 'Nº Centros Educativos', 
               'Nº Residencias Mayores', 'Nº Instituciones Religiosas y Militares']

Despoblacion[ColumnasNan] = Despoblacion[ColumnasNan].replace(np.nan, 0)

**Nº de extranjeros por municipio**: hay 772 nulos en hombres, 730 en mujeres y 466 en el total.
Además, si nos fijamos en la población total de estos municipios, esta está muy por debajo de la media.  
Entendemos los valores nulos como cero casos muestrales.

In [21]:
round(Despoblacion['Poblacion Total'].mean(), 2)

5834.07

In [22]:
NulosExtranjeros = Despoblacion[Despoblacion['Nº Total Extranjeros'].isnull()]
(NulosExtranjeros['Poblacion Total'] < 5834.07).sum()

466

In [23]:
ColumnasNan = ['Nº Hombres Extranjeros', 'Nº Mujeres Extranjeras', 'Nº Total Extranjeros']
Despoblacion[ColumnasNan] = Despoblacion[ColumnasNan].replace(np.nan, 0)

**Nº de personas otras comunidades autónomas**: Nº Hombres Otra CCAA hay 211 nulos, Nº Mujeres Otra CCAA hay 280 y en el total hay 96 nulos.  
Entendemos los valores nulos como cero casos muestrales.

In [24]:
ColumnasNan = ['Nº Hombres Otra CCAA', 'Nº Mujeres Otra CCAA', 'Nº Total Otra CCAA']
Despoblacion[ColumnasNan] = Despoblacion[ColumnasNan].replace(np.nan, 0)

**Nº de vehiculos**: 8 de los 9 valores nulos coinciden con que también son nulos los valores de la variable número de conductores.  
Asignamos el valor cero en ambos campos:

In [25]:
Despoblacion[Despoblacion['Nº Vehiculos'].isnull()]

Nombre CCAA Codigo CCAA Nombre Provincia Codigo Provincia  \
749       Extremadura          11          Badajoz               06   
2145          Galicia          12        Coruña, A               15   
4678          Navarra          15          Navarra               31   
4824          Navarra          15          Navarra               31   
7266  Castilla y León          07       Valladolid               47   
7448  Castilla y León          07       Valladolid               47   
7460  Castilla y León          07       Valladolid               47   
7637  Castilla y León          07           Zamora               49   
7789  Castilla y León          07           Zamora               49   

                Nombre Municipio Codigo Municipio  Poblacion Total  \
749              Sancti-Spíritus            06118            243.0   
2145                     Cesuras            15026           2225.0   
4678                    Castejón            31070           4316.0   
4824                        Sada            31212            195.0   
7266                Campillo, El            47031            231.0   
7448  Villanueva de los Infantes            47221            123.0   
7460                   Zarza, La            47232            126.0   
7637                     Fonfría            49071            950.0   
7789                 Villaescusa            49241            299.0   

      Poblacion Hombre  Poblacion Mujer  Superficie Km2  Densidad Poblacion  \
749              131.0            112.0           33.56                7.24   
2145            1090.0           1135.0           79.46               28.00   
4678            2175.0           2141.0           18.07              238.85   
4824             103.0             92.0           12.65               15.42   
7266             129.0            102.0           32.80                7.04   
7448              70.0             53.0           19.17                6.42   
7460              71.0             55.0           24.58                5.13   
7637             482.0            468.0          132.89                7.15   
7789             154.0            145.0           43.09                6.94   

      Altitud Municipios  Nacimientos  Matrimonios  Defunciones  \
749                485.0          1.0          0.0          7.0   
2145               231.0          3.0          9.0         35.0   
4678               277.0         51.0         16.0         41.0   
4824               510.0          1.0          0.0          4.0   
7266               744.0          2.0          1.0          3.0   
7448               745.0          1.0          0.0          0.0   
7460               757.0          2.0          0.0          1.0   
7637               804.0          0.0          1.0         15.0   
7789               827.0          0.0          0.0          9.0   

      Total Regimen SS  Total Contratos  Contratos Hombres  Contratos Mujeres  \
749               33.0             10.0                5.0                5.0   
2145             325.0             67.0               40.0               27.0   
4678            1151.0            947.0              519.0              428.0   
4824              53.0             31.0               21.0               10.0   
7266              28.0             24.0               24.0                0.0   
7448              25.0             11.0                9.0                2.0   
7460              27.0             16.0               13.0                3.0   
7637             120.0             41.0               32.0                9.0   
7789              55.0             24.0               14.0               10.0   

      Total Paro Registrado  Parados Hombres  Parados Mujeres  \
749                   277.0            160.0            117.0   
2145                 1396.0            940.0            456.0   
4678                 5098.0           3081.0           2017.0   
4824                   30.0             29.0              1.0   


In [26]:
Despoblacion[['Nº Vehiculos', 'Nº Conductores']] \
.loc[((Despoblacion['Nº Vehiculos'].isnull()) & (Despoblacion['Nº Conductores'].isnull()))] = 0

Queda un municipio con el número de vehículos cero y 2 conductores. Como la población es muy pequeña, 230 habitantes, asignamos el valor 0:

In [27]:
Despoblacion['Nº Vehiculos'] = Despoblacion['Nº Vehiculos'].replace(np.nan, 0)

**Nº condutores**: Hay 3 municipios con valores nulos y con nº de vehículos: 579, 2 y 6. En estos dos últimos asumimos que el número de conductores es cero. Para el municipio de 579 vehículos se ha buscado y encontrado el nº de conductores en 2012: 351. Sustituimos el nulo por dicho valor.  
Para los nulos con cero número de vehículos, sustituimos el valor por cero.

In [28]:
Despoblacion[Despoblacion['Nº Conductores'].isnull()]

Nombre CCAA Codigo CCAA Nombre Provincia Codigo Provincia  \
2145          Galicia          12        Coruña, A               15   
4626          Navarra          15          Navarra               31   
4678          Navarra          15          Navarra               31   
4824          Navarra          15          Navarra               31   
6286  Castilla y León          07            Soria               42   
6433         Cataluña          09        Tarragona               43   
7266  Castilla y León          07       Valladolid               47   
7448  Castilla y León          07       Valladolid               47   
7460  Castilla y León          07       Valladolid               47   
7637  Castilla y León          07           Zamora               49   
7789  Castilla y León          07           Zamora               49   

                Nombre Municipio Codigo Municipio  Poblacion Total  \
2145                     Cesuras            15026           2225.0   
4626                     Arantza            31022            644.0   
4678                    Castejón            31070           4316.0   
4824                        Sada            31212            195.0   
6286                    Rebollar            42151             44.0   
6433                   Molar, El            43085            305.0   
7266                Campillo, El            47031            231.0   
7448  Villanueva de los Infantes            47221            123.0   
7460                   Zarza, La            47232            126.0   
7637                     Fonfría            49071            950.0   
7789                 Villaescusa            49241            299.0   

      Poblacion Hombre  Poblacion Mujer  Superficie Km2  Densidad Poblacion  \
2145            1090.0           1135.0           79.46               28.00   
4626             354.0            290.0           31.71               20.31   
4678            2175.0           2141.0           18.07              238.85   
4824             103.0             92.0           12.65               15.42   
6286              22.0             22.0           10.38                4.24   
6433             166.0            139.0           22.91               13.31   
7266             129.0            102.0           32.80                7.04   
7448              70.0             53.0           19.17                6.42   
7460              71.0             55.0           24.58                5.13   
7637             482.0            468.0          132.89                7.15   
7789             154.0            145.0           43.09                6.94   

      Altitud Municipios  Nacimientos  Matrimonios  Defunciones  \
2145               231.0          3.0          9.0         35.0   
4626               270.0          9.0          2.0          6.0   
4678               277.0         51.0         16.0         41.0   
4824               510.0          1.0          0.0          4.0   
6286              1134.0          0.0          0.0          2.0   
6433               209.0          2.0          0.0          5.0   
7266               744.0          2.0          1.0          3.0   
7448               745.0          1.0          0.0          0.0   
7460               757.0          2.0          0.0          1.0   
7637               804.0          0.0          1.0         15.0   
7789               827.0          0.0          0.0          9.0   

      Total Regimen SS  Total Contratos  Contratos Hombres  Contratos Mujeres  \
2145             325.0             67.0               40.0               27.0   
4626              97.0             54.0               30.0               24.0   
4678            1151.0            947.0              519.0              428.0   
4824              53.0             31.0               21.0               10.0   
6286               7.0              3.0                1.0                2.0   
6433              61.0             19.0                9.0               10.

In [29]:
Despoblacion.loc[4626, 'Nº Conductores'] = 351
Despoblacion['Nº Conductores'] = Despoblacion['Nº Conductores'].replace(np.nan, 0)

**Total impuestos (Miles de €)**: Son 5 registros con valores nulos. Buscando estos valores en el ministerio de hacienda, el valor que les asignan es cero. Sustituimos los nulos por dicho valor. 

In [30]:
Despoblacion[Despoblacion['Total impuestos (Miles de €)'].isnull()]

Nombre CCAA Codigo CCAA Nombre Provincia Codigo Provincia  \
4     Pais Vasco          16      Araba/Álava               01   
6541      Aragón          02           Teruel               44   
6545      Aragón          02           Teruel               44   
7834      Aragón          02         Zaragoza               50   
7835      Aragón          02         Zaragoza               50   

        Nombre Municipio Codigo Municipio  Poblacion Total  Poblacion Hombre  \
4                Armiñón            01006            229.0             112.0   
6541  Alcalá de la Selva            44012            516.0             274.0   
6545              Aliaga            44017            382.0             214.0   
7834   Alcalá de Moncayo            50014            205.0             118.0   
7835  Alconchel de Ariza            50015             96.0              55.0   

      Poblacion Mujer  Superficie Km2  Densidad Poblacion  Altitud Municipios  \
4               117.0           12.97               17.66               467.0   
6541            242.0          104.95                4.92              1403.0   
6545            168.0          193.08                1.98              1105.0   
7834             87.0           13.76               14.90               763.0   
7835             41.0           34.85                2.75               893.0   

      Nacimientos  Matrimonios  Defunciones  Total Regimen SS  \
4             2.0          0.0          3.0             137.0   
6541          2.0          1.0          2.0             170.0   
6545          2.0          0.0         14.0              96.0   
7834          2.0          0.0          1.0              22.0   
7835          0.0          0.0          1.0              21.0   

      Total Contratos  Contratos Hombres  Contratos Mujeres  \
4               125.0               90.0               35.0   
6541            122.0               66.0               56.0   
6545             40.0               17.0               23.0   
7834             23.0               13.0               10.0   
7835             11.0                6.0                5.0   

      Total Paro Registrado  Parados Hombres  Parados Mujeres  \
4                     139.0             60.0             79.0   
6541                  362.0            195.0            167.0   
6545                  146.0             60.0             86.0   
7834                  158.0             81.0             77.0   
7835                   18.0             12.0              6.0   

      Nº Establecimientos Agricultura  Nº Establecimientos Industria  \
4                                 0.0                            0.0   
6541                              0.0                            1.0   
6545                              3.0                            3.0   
7834                              0.0                            0.0   
7835                              1.0                            0.0   

      Nº Establecimientos Construccion  Nº Establecimientos Servicios  \
4                                  0.0                            0.0   
6541                               5.0                           29.0   
6545                               3.0                           10.0   
7834                               3.0                            3.0   
7835                               0.0                            2.0   

      Nº Centros Sanitarios  Nº Centros Educativos  \
4                       0.0                    2.0   
6541                    1.0                    1.0   
6545                    3.0                    1.0   
7834                    1.0                    0.0   
7835                    1.0                    0.0   

      Número de transacciones anuales 2011  Nº Residencias Mayores  \
4                                      7.0                     0.0   
6541                                  27.0                     0.0   
6545                                   3.0                     0.0   
78

In [31]:
Despoblacion['Total impuestos (Miles de €)'] = Despoblacion['Total impuestos (Miles de €)'].replace(np.nan, 0)

**Superficie (Ha) Agricolas, Nº Explotaciones Agrícolas, Nº Explotaciones Ganaderas**: Los nulos son municipios en los que no hay actividad.  
Sustituimos los nulos por ceros.

In [32]:
Despoblacion[['Superficie (Ha) Agricolas', 'Nº Explotaciones Agrícolas', 'Nº Explotaciones Ganaderas' ]] = \
Despoblacion[['Superficie (Ha) Agricolas', 'Nº Explotaciones Agrícolas', 'Nº Explotaciones Ganaderas' ]].\
replace(np.nan, 0)

**Edad Media Edificios y número de establecimientos**: Hay 522 y 517 valores nulos respectivamente. 
En este caso son registros nulos de los que no disponemos de sus valores.  
Si decidimos eliminarlos tendremos el siguiente inconveniente:

In [33]:
Nulos = Despoblacion[Despoblacion['Edad Media Edificios'].isnull()]
Nulos['Nombre CCAA'].unique()

array(['Pais Vasco', 'Castilla y León', 'Galicia', 'Aragón', 'Navarra'],
      dtype=object)

In [34]:
print(Nulos['Codigo CCAA'].value_counts()['16']) #Codigo CCAA Pais Vasco
print(Nulos['Codigo CCAA'].value_counts()['15']) #Codigo CCAA Navarra

248
271


Pais Vasco y Navarra tienen 251 y 272 municipios respectivamente. Si eliminamos los nulos de Edad Media Edificios y número de establecimientos, nos quedamos prácticamente sin registros de estas dos CCAA.  
Por tanto, se decide eliminar estas dos variables del DataFrame.

In [35]:
Columnas = ['Edad Media Edificios', 'Uso Industrial', 'Uso Residelcial', 'Uso Terciario', 'Otros Usos']
Despoblacion = Despoblacion.drop(Columnas, axis = 1)

Eliminados todos los nulos:

In [36]:
Despoblacion.isnull().sum().sum()

0

DataFrame definitivo:

In [37]:
Despoblacion.head()

Nombre CCAA Codigo CCAA Nombre Provincia Codigo Provincia  Nombre Municipio  \
0  Pais Vasco          16      Araba/Álava               01  Alegría-Dulantzi   
1  Pais Vasco          16      Araba/Álava               01           Amurrio   
2  Pais Vasco          16      Araba/Álava               01             Añana   
3  Pais Vasco          16      Araba/Álava               01           Aramaio   
4  Pais Vasco          16      Araba/Álava               01           Armiñón   

  Codigo Municipio  Poblacion Total  Poblacion Hombre  Poblacion Mujer  \
0            01001           2803.0            1466.0           1337.0   
1            01002          10014.0            5030.0           4984.0   
2            01049            168.0              86.0             82.0   
3            01003           1496.0             784.0            712.0   
4            01006            229.0             112.0            117.0   

   Superficie Km2  Densidad Poblacion  Altitud Municipios  Nacimientos  \
0           19.95              140.50               568.0         37.0   
1           96.30              103.99               219.0        104.0   
2           21.92                7.66               574.0          1.0   
3           73.02               20.49               333.0         13.0   
4           12.97               17.66               467.0          2.0   

   Matrimonios  Defunciones  Total Regimen SS  Total Contratos  \
0         18.0         14.0             784.0            617.0   
1         37.0         88.0            4397.0           5861.0   
2          0.0          7.0              30.0             20.0   
3          3.0         10.0             224.0            141.0   
4          0.0          3.0             137.0            125.0   

   Contratos Hombres  Contratos Mujeres  Total Paro Registrado  \
0              453.0              164.0                 1992.0   
1             4843.0             1018.0                 8393.0   
2               14.0                6.0                   50.0   
3               88.0               53.0                  322.0   
4               90.0               35.0                  139.0   

   Parados Hombres  Parados Mujeres  Nº Establecimientos Agricultura  \
0            935.0           1057.0                              0.0   
1           3742.0           4651.0                              0.0   
2             16.0             34.0                              0.0   
3            177.0            145.0                              0.0   
4             60.0             79.0                              0.0   

   Nº Establecimientos Industria  Nº Establecimientos Construccion  \
0                            0.0                               0.0   
1                            0.0                               0.0   
2                            0.0                               0.0   
3                            0.0                               0.0   
4                            0.0                               0.0   

   Nº Establecimientos Servicios  Nº Centros Sanitarios  \
0                            0.0                    1.0   
1                            0.0                    2.0   
2                            0.0                    1.0   
3                            0.0                    1.0   
4                            0.0                    0.0   

   Nº Centros Educativos  Número de transacciones anuales 2011  \
0                    4.0                                  25.0   
1                    9.0                                 185.0   
2                    0.0                                   0.0   
3                    2.0                                   8.0   
4                    2.0                                   7.0   

   Nº Residencias Mayores  Nº Instituciones Religiosas y Militares  \
0                     1.0                                      1.0   
1                     3.0                                      0.0   
2              

In [38]:
Despoblacion.shape

(8077, 75)

In [39]:
Despoblacion.to_csv('Despoblacion_11.csv', header = True, index = False)